src/odemis/gui/cont/project.py

```diff
@@ -336,100 +356,192 @@ def _on_text(self, evt):
+                self._tab_data.main.multibeam.resolution.value[1] * self._tab_data.main.multibeam.pixelSize.value[1])
+        # zoom in on ROC; add some space around ROC (factor 5 defines zoom level)
+        cnvs.fit_to_bbox([xmin - 5 * size[0], ymin - 5 * size[1], xmax + 5 * size[0], ymax + 5 * size[1]])
 
-    def _on_coordinates(self, coords):
-        # Purely for logging
-        logging.debug("ROC '%s' coordinates changed to %s.", self.model.name.value, coords)
+    def _on_coordinates(self, coordinates):
+        """
+        Called when the coordinates of a region of calibration (ROC) object are changed.
+        This can be by either adding, removing or dragging a ROC.
+        :param coordinates: (UNDEFINED_ROI or tuple of 4 floats) Coordinates of the region of calibration (l, t, r, b).
+        """
+        logging.debug("ROC '%s' coordinates changed to %s.", self.calib_model.name.value, coordinates)
+        if coordinates == acqstream.UNDEFINED_ROI:
+            # remove the ROC overlay in the viewport
+            if self.overlay is not None:
+                self._view_ctrl.viewports[0].canvas.remove_overlay(self.overlay)
+                self.overlay = None  # needed so it can be added again to viewport
+        else:
+            # add ROC overlay if not there yet (e.g. do not add when just moving the overlay)
+            if self.overlay is None:
+                if self.calib_prefix == "calib_3":
+                    self.overlay = self._view_ctrl.viewports[0].canvas.\
+                        add_calibration_overlay(self.calib_model.coordinates,
+                                                self.calib_model.name.value,
+                                                colour="#00ff00")
+                else:  # use default color (orange)
+                    self.overlay = self._view_ctrl.viewports[0].\
+                        canvas.add_calibration_overlay(self.calib_model.coordinates, self.calib_model.name.value)
 
 
 class FastEMCalibrationRegionsController(object):
     """
     Listens to the calibration buttons and creates the FastEMROCControllers accordingly.
     """
 
-    def __init__(self, tab_data, calibration_bar, view_ctrl):
+    def __init__(self, tab_data, calibration_panel, view_ctrl, calib_prefix):
         """
-        tab_data (FastEMAcquisitionGUIData): tab data model
-        calibration_bar (FastEMCalibrationBar): main calibration panel
-        view_ctrl (FastEMAcquisitionViewport): viewport controller
+        :param tab_data (FastEMAcquisitionGUIData): The tab data model.
+        :param calibration_panel (FastEMCalibrationPanelHeader): The main calibration panel including the 9 regions
+                of calibration (ROC) buttons, the calibrate button, the gauge and the label.
+        :param view_ctrl (FastEMAcquisitionViewport): The viewport controller. TODO replace with just view
```

Comment:
It seems that it would simplify quite a few lines, and seems rather straightforward to do... could you do it now?